src/Adapter/Presenter/Product/ProductLazyArray.php

```diff
@@ -570,18 +595,59 @@ private function shouldShowPrice(
      *
      * @param array $product
      *
-     * @return mixed
+     * @return bool
      */
-    private function shouldShowAddToCartButton($product)
+    private function shouldShowAddToCartButton(array $product): bool
     {
         return (bool) $product['available_for_order'];
     }
 
-    private function fillImages(
-        ProductPresentationSettings $settings,
-        array $product,
-        Language $language
-    ) {
+    /**
+     * @param array $product
+     *
+     * @return bool
+     */
+    private function shouldShowOutOfStockLabel(array $product): bool
+    {
+        if (!$this->configuration->getBoolean('PS_SHOW_LABEL_OOS_LISTING_PAGES')) {
+            return false;
+        }
+
+        // Displayed only if the order of out of stock product is denied.
+        if ($product['out_of_stock'] == OutOfStockType::OUT_OF_STOCK_AVAILABLE
+            || (
+                $product['out_of_stock'] == OutOfStockType::OUT_OF_STOCK_DEFAULT
+                && $this->configuration->getBoolean('PS_ORDER_OUT_OF_STOCK')
+            )) {
+            return false;
+        }
+
+        if ($product['id_product_attribute']) {
+            // Displayed only if all combinations are out of stock (stock is <= 0)
+            $product = new Product((int) $product['id_product']);
+            if (empty($product->id)) {
+                return false;
+            }
+
+            foreach ($product->getAttributesResume($this->language->getId()) as $combination) {
```

Comment:
`Product::getAttributeCombinations` seems to be a little lighter, what do you think?